include/flang/Semantics/tools.h

```diff
@@ -269,6 +269,28 @@ template<typename T> std::optional<std::int64_t> GetIntValue(const T &x) {
   }
 }
 
+template<typename T> std::optional<bool> GetBoolValue(const T &x) {
+  if (const auto *expr{GetExpr(x)}) {
+    return evaluate::ToBool(*expr);
+  } else {
+    return std::nullopt;
+  }
+}
+
+template<typename S = std::string, typename T>
+std::optional<S> GetStringValue(const T &x) {
+  if (const auto *expr{GetExpr(x)}) {
+    if (const auto *charExpr{
+            evaluate::UnwrapExpr<evaluate::Expr<evaluate::SomeCharacter>>(
+                *expr)}) {
+      return ToString(
+          std::get<evaluate::Expr<evaluate::Type<TypeCategory::Character,
```

Comment:
This will crash if the expression is a character expression of a distinct kind.  It also assumes that the kind code values for `CHARACTER` are the same as the `sizeof` values for various C++ string elements.

Reply:
f18 supports 1, 2, and 4 character kinds. My understanding was, these are made available to encode UTF-8, UTF 16, UTF32 strings. But you mentioned there is nothing like UTF-16 and UTF-32.
Could you please elaborate on the supported kinds of character in f18 and its usage?
This will clarify most of our doubts.
I will remove the dependency on C++ string kinds for this function. 
This will again lead to 3 different functions for each kind, this was my earlier implementation.

Reply:
The following is what I could glean from the code. This might be stating the obvious but just thought will share.

From function SelectedCharKind in lib/Evaluate/type.cpp we can see that,
ascii/default = kind 1
ucs-2 = kind 2
iso_10646/ucs-4 = kind 4

In lib/Parser/characters.cpp we can see functions for encoding and decoding. For e.g. encoding strings,
```
template std::string EncodeString<Encoding::LATIN_1, std::string>(
    const std::string &);
template std::string EncodeString<Encoding::UTF_8, std::u16string>(
    const std::u16string &);
template std::string EncodeString<Encoding::UTF_8, std::u32string>(
    const std::u32string &);
```
 
In functionExpressionAnalyzer::AnalyzeString in lib/Semantics/expression.cpp we can see the relation between kinds and Encoding/Decoding.
```
  switch (kind) {
  case 1:
    return AsGenericExpr(Constant<Type<TypeCategory::Character, 1>>{
        parser::DecodeString<std::string, parser::Encoding::LATIN_1>(
            string, true)});
  case 2:
    return AsGenericExpr(Constant<Type<TypeCategory::Character, 2>>{
        parser::DecodeString<std::u16string, parser::Encoding::UTF_8>(
            string, true)});
  case 4:
    return AsGenericExpr(Constant<Type<TypeCategory::Character, 4>>{
        parser::DecodeString<std::u32string, parser::Encoding::UTF_8>(
            string, true)});
  default: CRASH_NO_CASE;
  }
```

Reply:
Thanks for sharing the details Kiran.